File: backend/app/visualization/spec.py
"""The canonical VisualizationSpec. This is the single source of truth for a chart's state --
everything downstream (validation, renderers, the studio, and eventually the Phase 2 AI
copilot) reads and writes this shape, never ad-hoc chart config. It is JSON-serializable so it
can be stored verbatim in VisualizationVersion.spec and diffed across versions.
"""

from __future__ import annotations

from enum import StrEnum

from pydantic import BaseModel, Field


class EncodingType(StrEnum):
    QUANTITATIVE = "quantitative"
    ORDINAL = "ordinal"
    NOMINAL = "nominal"
    TEMPORAL = "temporal"


class Aggregation(StrEnum):
    NONE = "none"
    SUM = "sum"
    MEAN = "mean"
    MEDIAN = "median"
    COUNT = "count"
    MIN = "min"
    MAX = "max"


class Encoding(BaseModel):
    field: str
    type: EncodingType
    aggregation: Aggregation = Aggregation.NONE
    label: str | None = None
    format: str | None = None
    """Number/date format string, e.g. '.1%' or '%b %Y' -- rendering hint only."""


class Encodings(BaseModel):
    x: Encoding | None = None
    y: Encoding | None = None
    color: Encoding | None = None
    size: Encoding | None = None
    detail: Encoding | None = None
    """Extra grouping field with no visual channel of its own (e.g. facet key)."""

    # --- Range / interval channels ---
    x2: Encoding | None = None
    """End of an x-range. Required by Gantt (task start -> end); the bar spans x..x2."""
    y2: Encoding | None = None
    """End of a y-range (e.g. an explicit band). Set automatically by the waterfall compiler."""

    # --- Second measure (dual-axis combination charts) ---
    measure2: Encoding | None = None
    """A second, independently-scaled measure. Used by line_column: `y` renders as columns,
    `measure2` as an overlaid line on its own axis."""

    # --- Financial OHLC channels ---
    open: Encoding | None = None
    high: Encoding | None = None
    low: Encoding | None = None
    close: Encoding | None = None
    """Open/high/low/close for candlestick and OHLC charts. All four are required together;
    `x` carries the period. Validated as a set, never partially."""


class FilterOperator(StrEnum):
    EQ = "eq"
    NEQ = "neq"
    GT = "gt"
    GTE = "gte"
    LT = "lt"
    LTE = "lte"
    IN = "in"
    NOT_NULL = "not_null"


class Filter(BaseModel):
    id: str
    field: str
    operator: FilterOperator
    value: str | float | list[str] | list[float] | None = None


class SortSpec(BaseModel):
    field: str
    descending: bool = False


class AnnotationType(StrEnum):
    CALLOUT = "callout"
    REFERENCE_LINE = "reference_line"
    HIGHLIGHTED_REGION = "highlighted_region"
    LABEL = "label"
    SOURCE_NOTE = "source_note"


class Annotation(BaseModel):
    id: str
    type: AnnotationType
    text: str
    target_field: str | None = None
    target_value: str | float | None = None
    """What the annotation points at, e.g. a specific x-value for a reference_line."""


class Typography(BaseModel):
    title: str | None = None
    subtitle: str | None = None
    source_note: str | None = None


class Layout(BaseModel):
    width: int | None = None
    height: int | None = None
    show_legend: bool = True
    show_grid: bool = True


class VisualizationMetadata(BaseModel):
    dataset_id: str
    dataset_version_id: str
    story_id: str | None = None
    generated_by: str = "deterministic"
    """Provenance: 'deterministic' (Story pipeline, no LLM), 'gemini' (AI-proposed chart, every
    field still re-validated against the real schema before this spec exists), or 'user' (manual
    studio edit)."""
    reasoning: str | None = None
    """Gemini's own one-sentence justification for this chart/field pairing, when
    generated_by='gemini' -- never shown as a claim of fact, just the model's stated reasoning."""


class VisualizationSpec(BaseModel):
    """Versioned, validated, renderer-agnostic. Never contains executable code -- only
    declarative field/encoding/style references the renderer interprets."""

    chart_type: str
    encoding: Encodings = Field(default_factory=Encodings)
    transformations: list[str] = Field(default_factory=list)
    """Named transform references (e.g. 'coerce_numeric:revenue') applied before rendering --
    not arbitrary code."""
    filters: list[Filter] = Field(default_factory=list)
    sort: SortSpec | None = None
    annotations: list[Annotation] = Field(default_factory=list)
    theme: str = "minimal"
    typography: Typography = Field(default_factory=Typography)
    layout: Layout = Field(default_factory=Layout)
    metadata: VisualizationMetadata

File: frontend/src/lib/visualization/spec.ts
/**
 * TypeScript mirror of backend/app/visualization/spec.py::VisualizationSpec.
 * Keep these two in sync by hand until a codegen step exists (no frontend
 * visualization UI existed when the Python side was written, so this mirror
 * is being added now, on first actual frontend use).
 */

export type EncodingType = "quantitative" | "ordinal" | "nominal" | "temporal";

export type Aggregation = "none" | "sum" | "mean" | "median" | "count" | "min" | "max";

export interface Encoding {
  field: string;
  type: EncodingType;
  aggregation?: Aggregation;
  label?: string | null;
  format?: string | null;
}

export interface Encodings {
  x?: Encoding | null;
  y?: Encoding | null;
  color?: Encoding | null;
  size?: Encoding | null;
  detail?: Encoding | null;

  /** End of an x-range. Required by Gantt (task start -> end); the bar spans x..x2. */
  x2?: Encoding | null;
  /** End of a y-range. Set automatically by the waterfall compiler. */
  y2?: Encoding | null;

  /** A second, independently-scaled measure -- line_column renders `y` as columns and this
   * as an overlaid line on its own axis. */
  measure2?: Encoding | null;

  /** Open/high/low/close for candlestick and OHLC charts; all four required together. */
  open?: Encoding | null;
  high?: Encoding | null;
  low?: Encoding | null;
  close?: Encoding | null;
}

export type FilterOperator = "eq" | "neq" | "gt" | "gte" | "lt" | "lte" | "in" | "not_null";

export interface VizFilter {
  field: string;
  operator: FilterOperator;
  value?: string | number | string[] | number[] | null;
}

export interface SortSpec {
  field: string;
  descending: boolean;
}

export type AnnotationType =
  | "callout"
  | "reference_line"
  | "highlighted_region"
  | "label"
  | "source_note";

export interface Annotation {
  id: string;
  type: AnnotationType;
  text: string;
  target_field?: string | null;
  target_value?: string | number | null;
}

export interface Typography {
  title?: string | null;
  subtitle?: string | null;
  source_note?: string | null;
}

export interface Layout {
  width?: number | null;
  height?: number | null;
  show_legend: boolean;
  show_grid: boolean;
}

export interface VisualizationMetadata {
  dataset_id: string;
  dataset_version_id: string;
  story_id?: string | null;
}

export interface VisualizationSpec {
  chart_type: string;
  encoding: Encodings;
  transformations: string[];
  filters: VizFilter[];
  sort?: SortSpec | null;
  annotations: Annotation[];
  theme: string;
  typography: Typography;
  layout: Layout;
  metadata: VisualizationMetadata;
}
